### PR TITLE
Implemented the Record API.

### PR DESCRIPTION
**Pull request by Cybermaxke**
[API](https://github.com/SpongePowered/SpongeAPI/pull/1609) | **Common**

Implements the features added by the API. Also implements the `getDefaultProperty` method in the `ItemStack`.

### CODE REVIEW

File: src/main/java/org/spongepowered/common/mixin/core/entity/player/MixinEntityPlayerMP.java [hunk 1]
```diff
@@ -687,6 +693,23 @@ public void playSound(SoundType sound, SoundCategory category, Vector3d position
...
+
+    @Override
+    public void stopRecord(Vector3i position) {
+        playRecord0(null, position);
+    }
+
+    private void playRecord0(@Nullable RecordType recordType, Vector3i position) {
+        checkNotNull(position, "position");
+        final BlockPos pos = new BlockPos(position.getX(), position.getY(), position.getZ());
+        this.connection.sendPacket(new SPacketEffect(1010, pos,
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/mixin/core/entity/player/MixinEntityPlayerMP.java]: Extract `1010` to a constant somewhere, and use it here and any other occurrences.

File: src/main/java/org/spongepowered/common/mixin/core/entity/player/MixinEntityPlayerMP.java [hunk 2]
```diff
@@ -687,6 +693,23 @@ public void playSound(SoundType sound, SoundCategory category, Vector3d position
...
+    }
+
+    @Override
+    public void stopRecord(Vector3i position) {
+        playRecord0(null, position);
+    }
+
+    private void playRecord0(@Nullable RecordType recordType, Vector3i position) {
+        checkNotNull(position, "position");
+        final BlockPos pos = new BlockPos(position.getX(), position.getY(), position.getZ());
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/mixin/core/entity/player/MixinEntityPlayerMP.java]: There's a method in `VecHelper` to convert.

File: src/main/java/org/spongepowered/common/mixin/core/world/MixinWorldServer.java [hunk 3]
```diff
@@ -2167,6 +2170,23 @@ public void spawnParticles(ParticleEffect particleEffect, Vector3d position, int
...
+    public void stopRecord(Vector3i position) {
+        playRecord0(null, position);
+    }
+
+    private void playRecord0(@Nullable RecordType recordType, Vector3i position) {
+        checkNotNull(position, "position");
+        final PlayerList playerList = this.mcServer.getPlayerList();
+        final BlockPos pos = new BlockPos(position.getX(), position.getY(), position.getZ());
+        playerList.sendPacketToAllPlayersInDimension(new SPacketEffect(
+                1010, pos, recordType == null ? 0 : ((SpongeRecordType) recordType).getInternalId(), false), getDimensionId());
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/mixin/core/world/MixinWorldServer.java]: See note about `1010` above.

**Review by kashike — CHANGES_REQUESTED** [review]

**Comment by Meronat** [comment]
Rebasing this on bleeding and running this caused no errors/test issues and the test plugin worked as intended for me. 

The test plugin uses the deprecated `getItem()` instead of `getType()` though.

File: testplugins/src/main/java/org/spongepowered/test/RecordTest.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+import org.spongepowered.api.item.inventory.ItemStack;
+import org.spongepowered.api.plugin.Plugin;
+
+import java.util.Optional;
+
+@Plugin(id = "record_test", name = "Record Test", version = "Right click to start/stop a record at a position.")
+public class RecordTest {
+
+    @Listener
+    public void onInit(GameInitializationEvent event) {
```
Comment (parlough) [inline testplugins/src/main/java/org/spongepowered/test/RecordTest.java:44]: Can remove this listener, you don't use it.

File: testplugins/src/main/java/org/spongepowered/test/RecordTest.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    @Listener
+    public void onPlayerInteract(InteractItemEvent.Secondary event, @Root Player player) {
+        final ItemStack itemStack = player.getItemInHand(HandTypes.MAIN_HAND).orElse(null);
+        if (itemStack != null) {
+            final Optional<RecordProperty> optRecord = itemStack.getProperty(RecordProperty.class);
+            if (optRecord.isPresent()) {
+                player.playRecord(player.getLocation().getPosition().toInt(), optRecord.get().getValue());
+            } else if (itemStack.getItem() == ItemTypes.SPONGE) {
```
Comment (parlough) [inline testplugins/src/main/java/org/spongepowered/test/RecordTest.java:54]: Switch to `itemStack.getType()`

**Review by parlough — APPROVED** [review]
Good addition, worked well for me after rebasing on bleeding.

**Pull request closed without merging**